### PR TITLE
test: more comprehensive username-password tests

### PR DESCRIPTION
**Pull request by danielgent**
This PR possibly begs more questions than it solves :clown_face:

### CODE REVIEW

File: integration-test/flows/password.spec.ts [hunk 1]
```diff
@@ -144,4 +144,101 @@ describe("password-flow", () => {
...
+        iss: "https://example.com/",
+        name: "Åkesson Þorsteinsson",
+        nickname: "Åkesson Þorsteinsson",
+        picture: "https://example.com/foo.png",
+      });
+    });
+  });
+  // TO TEST
+  // - login with existing user & password, but wrong password
+  // - login with non-existing user & password
```
Comment (danielgent) [inline integration-test/flows/password.spec.ts:242]: I should do these first two...

File: integration-test/flows/password.spec.ts [hunk 2]
```diff
@@ -143,5 +143,351 @@ describe("password-flow", () => {
         iss: "https://example.com/",
       });
     });
+    // I think I should change this test to expect rejections, and then skip it
+    // seems like something I should look into anyway!
+    it("should reject password signup for existing username-password user BUT it is not", async () => {
```
Comment (danielgent) [inline integration-test/flows/password.spec.ts:148]: :smile:  this test looks like an issue


I think this PR is large enough *BUT* I'll investigate this next _on a new PR_

These email-password routes actually exist so it's worth checking we can't do basic hacks like signing up an email-password user again :smile:

File: integration-test/flows/password.spec.ts [hunk 3]
```diff
@@ -143,5 +143,351 @@ describe("password-flow", () => {
...
+    // OR do we add more fixtures in the initial setup? e.g. create and login with a code user
+    // Markus mentioned having the fixtures MUCH more populated. Probably a good idea.
+    // OR at least we seed the database directly
+    // We would then check that account linking happens (not implemented)
+    // currently we would just return the code user...
+
+    // TO TEST--------------------------------------------------------
+    // same username-password user but a different tenant
+  });
+  describe("Login with password", () => {
```
Comment (danielgent) [inline integration-test/flows/password.spec.ts:268]: These tests are all quick and fairly sensible I think. Just to sanity check we're not doing anything totally lax :skull_and_crossbones: :skull_and_crossbones: :skull_and_crossbones:

File: integration-test/flows/password.spec.ts [hunk 4]
```diff
@@ -143,5 +143,351 @@ describe("password-flow", () => {
...
+
+    // TO TEST--------------------------------------------------------
+    // should do what with registration signup for existing email (code) user?
+    // - create new code user - copy-paste code from code flow tests
+    // - need to go the whole way and actually sign in with the code user to check it works
+    // OR do we add more fixtures in the initial setup? e.g. create and login with a code user
+    // Markus mentioned having the fixtures MUCH more populated. Probably a good idea.
+    // OR at least we seed the database directly
+    // We would then check that account linking happens (not implemented)
+    // currently we would just return the code user...
```
Comment (danielgent) [inline integration-test/flows/password.spec.ts:263]: more things I can think of testing.  I've got other open PRs so it's probably worth me adding some sanity checks.

We're not allowing password signups in production of course _but_ they are exposed standard OAuth routes

File: integration-test/flows/password.spec.ts [hunk 5]
```diff
@@ -143,5 +143,351 @@ describe("password-flow", () => {
...
+    // I think I should change this test to expect rejections, and then skip it
+    // seems like something I should look into anyway!
+    it("should reject password signup for existing username-password user BUT it is not", async () => {
+      const password = "password";
+
+      // ------------------------------------
+      // create a user with username-password
+      // ------------------------------------
+
+      // where is the tenant passed here? Interesting...
```
Comment (markusahlstrand) [inline integration-test/flows/password.spec.ts:155]: The clientId is tied to a tenant right?

File: integration-test/flows/password.spec.ts [hunk 6]
```diff
@@ -143,5 +143,351 @@ describe("password-flow", () => {
...
+          },
+          method: "POST",
+          body: JSON.stringify({
+            email: "password-login-test@example.com",
+            password,
+          }),
+        },
+      );
+      // interesting - I didn't expect this! Seems like a bug?
+      expect(createUserResponseRepeated.status).toBe(201);
```
Comment (markusahlstrand) [inline integration-test/flows/password.spec.ts:235]: Yeah, I'm guessing that this should be... a 409? That would give away that this user exists.. Maybe auth0 returns a 500 here?

**Review by markusahlstrand — APPROVED** [review]
Looks good. We should fix this :)

File: integration-test/flows/password.spec.ts (hunk 5 above)
Comment (danielgent) [reply integration-test/flows/password.spec.ts:155]: *lol* yes oops. I thought I deleted this :smile:

Comment (danielgent) [reply integration-test/flows/password.spec.ts:155]: the `clientId` is in the URL of course!

Comment (danielgent) [reply integration-test/flows/password.spec.ts:155]:
```suggestion
```

File: integration-test/flows/password.spec.ts (hunk 6 above)
Comment (danielgent) [reply integration-test/flows/password.spec.ts:234]: I'll check on the next PR.

I haven't used these endpoints directly

**Pull request merged**